galsim/chromatic.py

```diff
@@ -1359,3 +1376,286 @@ def withScaledFlux(self, flux_ratio):
+
+    def withSED(self, sed='flat'):
+        """
+        Make a new InterpolatedChromaticObject with the requested SED.
+
+        @param sed    Either a galsim.SED object to assign to this object as its SED, or a string
+                      'flat' which means a flat SED (equal weight to all wavelengths).  'flat' is
+                      what should be used for PSFs, so they do not modify the flux of the objects
+                      with which they are being convolved. 
+                      [default: 'flat']
+        @return a new InterpolatedChromaticObject
+        """
+        new = self.copy()
+        if sed == 'flat':
+            new.SED = lambda w: 1.0
+        else:
+            new.SED = sed
+        return new
+
+    def evaluateAtWavelength(self, wave, force_eval=False):
+        """
+        Evaluate this InterpolatedChromaticObject at a particular wavelength, either using
+        interpolation or via direct calculation, depending on how the object was originally
+        instantiated.
+
+        @param wave     Wavelength in nanometers.
+
+        @returns the monochromatic object at the given wavelength, as an InterpolatedImage.
+        """
+        if self.waves is not None and not force_eval:
+            im, stepk, maxk = self._imageAtWavelength(wave)
+            return galsim.InterpolatedImage(im, _force_stepk=stepk, _force_maxk=maxk)
+        else:
+            return self.simpleEvaluateAtWavelength(wave)
+
+    def _imageAtWavelength(self, wave):
+        """
+        Get an image of the object at a particular wavelength, using linear interpolation between
+        the originally-stored images.
+        """
+        # First, some wavelength-related sanity checks.
+        if self.waves is None:
+            raise RuntimeError("Requested image at some wavelength when doing direct calculation!")
+        if wave < min(self.waves) or wave > max(self.waves):
+            raise RuntimeError("Requested wavelength %.1f is outside the allowed range:"
+                               " %.1f to %.1f nm"%(wave, min(self.waves), max(self.waves)))
+
+        # Figure out where the supplied wavelength is compared to the list of wavelengths on which
+        # images were originally tabulated.
+        lower_idx = np.searchsorted(self.waves, wave)-1
+        # There can be edge issues, so watch out for that:
+        if lower_idx < 0: lower_idx = 0
+        if lower_idx == len(self.waves)-1: lower_idx = len(self.waves)-1
+        frac = (wave-self.waves[lower_idx]) / (self.waves[lower_idx+1]-self.waves[lower_idx])
+
+        # Actually do the interpolation for the image, stepK, and maxK.
+        im = _linearInterp(self.ims, frac, lower_idx)
+        stepk = _linearInterp(self.stepK_vals, frac, lower_idx)
+        maxk = _linearInterp(self.maxK_vals, frac, lower_idx)
+        return self.SED(wave)*im, stepk, maxk
+
+    def drawImage(self, bandpass, force_eval=False, image=None, **kwargs):
```

Comment:
I just did this.  The good part:

- The code seems cleaner to me.
- If I run the test script described above, then for the 3 cases I mentioned, the time to render the PSF image with 23, 54, and 73 elements in ```bandpass.wave_list``` is typically 0.5, 0.5, and 0.9 seconds. (Instead of 0.5, 0.75, and 1.7 seconds.)  So the scaling with the number of elements is definitely shallower.  

The weird part:
- I determined that the step that is dominating the times is the step where the InterpolatedImage is drawn (at the very end of drawImage).  For the cases of 54 and 73 elements, that step takes ~0.37 and ~0.7 seconds.  I'm a little confused as to why it's taking longer in one case than in the other - I have no idea why there should be any difference.  The stepk and maxk values are a bit different, but not very.  Any thoughts?